Support to read and update wifi guest network settings

frimtec opened the pull request:
Dear @ties

Thanks a lot for the useful compal library.

The only thing I'm missing, is the possibility to update wifi guest network settings.
I implemented the feature and tested it with my compal modem.

It would be very nice if you could integrate this feature into your library.

If there are issues with my python style, please forgive me, this is my first python code ;-)

Best regards,
Markus

frimtec commented:
Sorry I have no clue why these tests are failing - sorry, my python knowledge gap. 
If you could give me a hint, I fix it.

ties commented on compal/__init__.py:
```diff
@@ -926,6 +927,204 @@ def turn_off(self, debug=False):
+        assert band in ("2g", "5g",)
+
+        all_interfaces = list()
+        interfaces = xml.iter("Interface" + ("" if band == "2g" else band.upper()))
```
In the else it is always "5G"? I would use "5G" there instead of `band.upper()`

(I probably did this another way somewere else myself)

frimtec replied:
Good point - is changed.

ties commented on compal/__init__.py:
```diff
@@ -926,6 +927,204 @@ def turn_off(self, debug=False):
+        all_interfaces = list()
+        interfaces = xml.iter("Interface" + ("" if band == "2g" else band.upper()))
+        for interface in interfaces:
+            def xml_value(attr, coherce=True):
```
If I read this correctly, only `interface` is used from the scope the function is defined in. This means that you could define this function on the class (for example `_xml_value(self, interface, attr, coherce=True)`) and pass in the interface instead of in the loop.

frimtec replied:
Your absolut right, I moved the function to class level.

ties commented on compal/models.py:
```diff
@@ -33,6 +33,27 @@ class RadioSettings:
+
+@dataclass
+class GuestNetworkSettings:
+    guest_networks_2g: list[InterfaceGuestNetworkSettings] = None
```
I think these need to be `List` to fix the warnings,
```suggestion
    guest_networks_2g: List[InterfaceGuestNetworkSettings] = None
```

Line 4 also needs to import it, such as with `from typing import List, Optional`

ties replied:
I checked it: If you change this the tests probably pass!

frimtec replied:
Thanks for this hint. 
I changed it accordingly - hope the tests will not fail anymore.

ties reviewed (changes requested):
Thanks for the contribution! I can not test it, but it looks like a useful feature. I liked the docstrings that are included.

The linter is a bit fiddly. I spotted one thing that it will fail on, let's see if fixing that one is enough to make it happy 👍 .

ties commented on compal/__init__.py:
```diff
@@ -926,6 +927,204 @@ def turn_off(self, debug=False):
+        start_time = time.time()
+        changes = ""
+        if debug:
+            print(f"\n--- WAITING FOR ROUTER TO UPDATE ---")
```
```suggestion
            print("\n--- WAITING FOR ROUTER TO UPDATE ---")
```

The linter only accepts string-interpolation-syntax (`f"..."`) if there is a variable being used. So `f"{total}"` is ok but `f"not using a variable"` would not be accepted.

frimtec replied:
Thanks, it is fixed.

ties commented on compal/__init__.py:
```diff
@@ -926,6 +927,204 @@ def turn_off(self, debug=False):
+        """
+        Method for updating the wifi guest network settings. Uses fun:308.
+        """
+        assert 0 <= interface_index < (len(new_guest_network_settings.guest_networks_2g) and
```
I think `(len(new_guest_network_settings.guest_networks_2g) and                                        len(new_guest_network_settings.guest_networks_5g))` probably does not do what you expect. It returns 0 if the first value is 0, otherwise it returns the second value.

I would write:
```
assert 0 <= interface_index and interface_index < max(len(new_guest_network_settings.guest_networks_2g), len(new_guest_network_settings.guest_networks_5g))
```

But please make sure that does what you intend

frimtec replied:
Oh no, what a pity, I must have been tired when I wrote this ;-)  
You are right, it was not what I intended to do.
It is fixed now.

ties commented:
I can also help you fix the linter issues, just let me know :)

frimtec commented:
Many thank for your valuable comments. I have changed all mentioned points.

But I have no idea why the tests still fail. 
I fixed all the reported linter errors,  but somehow the tests are still red :-(

Can you help?

Many thanks and reagrds,
Markus

ties commented:
I'll take a look tonight! :)

ties commented:
I ran an autoformat (with `black`) and it passed. Thanks for the contribution!

frimtec commented:
Many thanks for accepting my change.
